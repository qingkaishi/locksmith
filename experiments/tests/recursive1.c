#include <pthread.h>
#include <stdio.h>

int f() {return 0;}

struct foo {
  void (*bar)(struct foo* a);
};

struct foo a;

void b(struct foo *a) {
  a->bar = b;
  a->bar(a);
}

pthread_mutex_t l;

int i;

void * dostuff(void * z) {
  pthread_mutex_lock(&l);
  a.bar = b;
  a.bar(NULL);
  pthread_mutex_unlock(&l);
  return NULL;
}

int main() {
  pthread_t t1;
  int i = 1;
  pthread_mutex_init(&l,NULL);
  pthread_create(&t1, NULL, dostuff, NULL);
  pthread_create(&t1, NULL, dostuff, NULL);
}


