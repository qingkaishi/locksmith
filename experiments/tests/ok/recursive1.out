************************* STARTING *************************

shared: alloc.bar:recursive1.c:10
alloc.bar:recursive1.c:10 is protected by:
  concrete l:recursive1.c:32
  l:recursive1.c:17
*************************** DONE ***************************
